crmsh/corosync.py

```diff
@@ -669,215 +649,372 @@ def remove_qdevice_db(self):
+        # nodelist.node.nodeid__1 -> nodelist.node__1
+        return re.sub("(.*)\.\w+(__[0-9])", "\\1\\2", inner_key)
 
-class Parser(object):
-    def __init__(self, data):
-        self._tokens = list(corosync_tokenizer(data))
-
-    def find(self, name, start=0):
-        """Gets the index of the element with the given path"""
-        for i, t in enumerate(self._tokens[start:]):
-            if t.path == name:
-                return i + start
-        return -1
-
-    def find_bounds(self, name, start=0):
-        """find the (start, end) of the next instance of name found at start"""
-        i = self.find(name, start)
-        if i < 0:
-            return -1, -1
-        if self._tokens[i].token != _tBEGIN:
-            return i, i
-        e = i + 1
-        depth = 0
-        while e < len(self._tokens):
-            t = self._tokens[e]
-            if t.token == _tBEGIN:
-                depth += 1
-            if t.token == _tEND:
-                depth -= 1
-            if depth < 0:
-                break
-            e += 1
-        if e == len(self._tokens):
-            raise ValueError("Unclosed section")
-        return i, e
-
-    def get(self, path):
-        """Gets the value for the key (if any)"""
-        for t in self._tokens:
-            if t.token == _tVALUE and t.path == path:
-                return t.value
-        return None
+    @staticmethod
+    def _section_name(inner_key):
+        """
+        Extract section name from an inner key
+        """
+        # nodelist.node.nodeid__1 -> nodelist.node
+        return re.sub("(.*)\.\w+(__[0-9])", "\\1", inner_key)
 
-    def get_all(self, path):
-        """Returns all values matching path"""
-        ret = []
-        for t in self._tokens:
-            if t.token == _tVALUE and t.path == path:
-                ret.append(t.value)
-        return ret
+    @staticmethod
+    def _key_name(inner_key):
+        """
+        Extract key name from an inner key
+        """
+        # nodelist.node.nodeid__1 -> nodelist.node.nodeid
+        return re.sub("__[0-9]+", "", inner_key)
+
+    @property
+    def _inner_section_name_list(self):
+        """
+        Get list of inner section name
+        """
+        return [key for key, value in self._config_dict.items() if value == "{"]
+
+    @property
+    def _inner_section_end_list(self):
+        """
+        Get list of inner section end
+        """
+        return [key for key, value in self._config_dict.items() if value == "}"]
+
+    def _is_inner_section(self, inner_key):
+        """
+        Check if an inner key is an inner section name
+        """
+        return inner_key in self._inner_section_name_list
 
-    def all_paths(self):
-        """Returns all value paths"""
+    def _unused_inner_key(self, path):
+        """
+        Generate unused inner key
+        """
+        index = 0
+        key = self._inner_key(path, index)
+        while key in self._config_dict:
+            index += 1
```

Comment:
You should manage this `index` variable with the `enumerate` option.
https://www.tutorialspoint.com/enumerate-in-python